gazebo/gazebo.1.ronn
gazebo -- Run the Gazebo server and GUI.
=============================================

## SYNOPSIS

`gazebo` [options] <world_file>

## DESCRIPTION

Gazebo server runs simulation and handles commandline options, starts a Master, runs World update and sensor generation loops. This also starts the Gazebo GUI client in a separate process.

## OPTIONS

* -v, --version :
 Output version information.
* --verbose :
 Increase the messages written to the terminal.
* -h, --help :
 Produce this help message.
* -u, --pause :
 Start the server in a paused state.
* -e, --physics arg :
 Specify a physics engine (ode|bullet|kido|simbody).
* -p, --play arg :
 Play a log file.
* -r, --record :
 Record state data.
* --record_encoding arg (=zlib) :
 Compression encoding format for log data (zlib|bz2|txt).
* --record_path arg :
 Absolute path in which to store state data.
* --seed arg :
 Start with a given random number seed.
* --iters arg :
 Number of iterations to simulate.
* --minimal_comms :
 Reduce the TCP/IP traffic output by gazebo.
* -g, --gui-plugin arg :
 Load a GUI plugin.
* -s, --server-plugin arg :
 Load a server plugin.
* -o, --profile arg :
 Physics preset profile name from the options in the world file.


## AUTHOR
  Open Source Robotics Foundation

## COPYRIGHT 
  Copyright (C) 2012-2016 Open Source Robotics Foundation
       
  Licensed under the Apache License, Version 2.0 (the "License");
  you may not use this file except in compliance with the License.


debian/changelog
gazebo (7.3.0+dfsg-3) unstable; urgency=medium

  [ Jochen Sprickerhof ]
  * Add libcurl-dev as an alternative to simplify apts resolution

  [ Jose Luis Rivero ]
  * Include patch to fix tinyxml constant removal
    (Closes: #835427)

 -- Jose Luis Rivero <jrivero@osrfoundation.org>  Mon, 29 Aug 2016 00:44:02 +0000

gazebo (7.3.0+dfsg-2) unstable; urgency=medium

   * Patch to fix linking with gcc6 (Closes: #831176)

 -- Jose Luis Rivero <jrivero@osrfoundation.org>  Thu, 21 Jul 2016 15:31:57 +0000

gazebo (7.3.0+dfsg-1) unstable; urgency=medium

  * Explicitly use of external tinyxml copy
  * Imported Upstream version 7.3.0

 -- Jose Luis Rivero <jrivero@osrfoundation.org>  Thu, 14 Jul 2016 18:53:12 +0000

gazebo (7.0.0+dfsg-2) unstable; urgency=medium

  * [9916e19] Fix Breaks/Replaces dependecies. (Closes: #817189)
  * [e6f3533] Remove dbg-package.
  * [8549abf] Apply cme fix dpkg.

 -- Anton Gladky <gladk@debian.org>  Sat, 12 Mar 2016 20:56:24 +0100

gazebo (7.0.0+dfsg-1) unstable; urgency=medium

  * Need sdformat4 in gazebo7
    (Closes: #813001)
  * Add libqtwebkit dependency
  * Imported Upstream version 7.0.0
  * Migrate all package metadata to gazebo7
  * Remove patch. Circular dependencies problem is fixed upstream

 -- Jose Luis Rivero <jrivero@osrfoundation.org>  Tue, 26 Jan 2016 13:49:54 +0100

gazebo (6.5.0+dfsg-2) unstable; urgency=medium

  [ Jose Luis Rivero ]
  * [7a3cc42] Remove player as dependency. (Closes: #802032)

  [ Anton Gladky ]
  * [6f51db4] Replace deprecated FFmpeg API. (Closes: #803816)
  * [7c0b8f1] Apply cme fix dpkg-control.
  * [37f9b28] Add myself to uploaders.

 -- Anton Gladky <gladk@debian.org>  Tue, 10 Nov 2015 21:41:54 +0100

gazebo (6.5.0+dfsg-1) unstable; urgency=medium

  * Added filter to gdp.conf
  * Imported Upstream version 6.5.0
  * Migrate patch to use FindGtest from PCL lib
  * Removed patches merged upstream and update the reamining ones.
  * Change all gazebo5 package references to gazebo6
  * Added patch to use system gtest
  * Added patch to fix ciruclar linking problem in gz

 -- Jose Luis Rivero <jrivero@osrfoundation.org>  Mon, 26 Oct 2015 17:04:30 +0100

gazebo (5.0.1+dfsg-2.1) unstable; urgency=medium

  * Non-maintainer upload.
  * debian/patches/0010-bullet-2.83.patch: Apply upstream patch to fix build
    against bullet 2.83. (Closes: #792992)

 -- Sebastian Ramacher <sramacher@debian.org>  Mon, 27 Jul 2015 18:35:36 +0200

gazebo (5.0.1+dfsg-2) unstable; urgency=medium

  * [208ddf2] Disable robot-player-dev support (see #787007)
    (Closes: #787007)

 -- Jose Luis Rivero <jrivero@osrfoundation.org>  Fri, 29 May 2015 19:42:29 +0200

gazebo (5.0.1+dfsg-1) unstable; urgency=medium

  * Team upload.
  * Upload into unstable

 -- Anton Gladky <gladk@debian.org>  Sat, 09 May 2015 23:42:37 +0200

gazebo (5.0.1+dfsg-1~exp3) experimental; urgency=medium

  * [337a3cc] Fix problems with upgrade paths
    (Closes: #781651, #782104, #782105)

 -- Jose Luis Rivero <jrivero@osrfoundation.org>  Wed, 08 Apr 2015 02:02:08 +0200

gazebo (5.0.1+dfsg-1~exp2) experimental; urgency=medium

  * [99f894c] Backport from upstream a patch to solve GAZEBO_MODEL_PATH bug

 -- Jose Luis Rivero <jrivero@osrfoundation.org>  Wed, 25 Mar 2015 18:17:06 +0000

gazebo (5.0.1+dfsg-1~exp1) experimental; urgency=medium

  * [eacde0a] Fix patch to remove baemuk license
  * [cb7f2a6] Added the Ronn2Man license
  * [74d51bd] Remove old baemuk license
  * [3ad9ccc] Include missing license files
  * [59db7c2] Update ann files copyright
  * [5ee8de5] Added maiden copyright
  * [03f23fd] Update Dejavu font copyright. Include license
  * [da384e9] Include rtshader files license
  * [a28256d] Update OSRF copyright to 2015
  * [2c7763e] Remove wrong included license
  * [969bf4c] Remove maiden.ttf font. Not in use
  * [65d716c] Remove duplicate robot-player-dev
  * [a6c50d6] Fix all lintian warnings in copyright file
  * [f2a16d5] Added libgazebo6-dev transitional package
  * [4b24607] Imported Upstream version 5.0.1. (Closes: #766748)
  * [556803f] Update install files to the new version
  * [3231500] Rename -dev file to libgazebo6-dev
  * [b6b788a] Update control file to gazebo6
  * [922909e] Update all references to gazebo3 to gazebo6
  * [cb4df5e] Remove patches that are now in upstream code

 -- Jose Luis Rivero <jrivero@osrfoundation.org>  Thu, 12 Feb 2015 02:39:29 +0000

gazebo (3.0.0+dfsg-3) unstable; urgency=medium

  * Team upload.
  * [84c0ad8] Simplify d/rules. Fix install-command. (Closes: #766303)

 -- Anton Gladky <gladk@debian.org>  Thu, 13 Nov 2014 17:58:47 +0100

gazebo (3.0.0+dfsg-2) unstable; urgency=medium

  * Enable simbody support
  * Enable player support
  * Enable gdal support
  * Update standard version to 3.9.6
  * Added copyright for fcl (thanks to Thorsten Alteholz)
  * Modify skyx copyright to be LGPLv2+
  * Missing command to get DEB_HOST_ARCH in rules file
  * Fix font file linking (Closes: #766303)

 -- Jose Luis Rivero <jrivero@osrfoundation.org>  Tue, 28 Oct 2014 22:44:07 +0000

gazebo (3.0.0+dfsg-1) unstable; urgency=medium

  * Initial release. (Closes: #705440)

 -- Jose Luis Rivero <jrivero@osrfoundation.org>  Tue, 17 Jun 2014 14:50:35 +0200


gazebo/gazebo.cc
/*
 * Copyright (C) 2012-2015 Open Source Robotics Foundation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */
#include <vector>
#include <boost/thread/mutex.hpp>
#include <sdf/sdf.hh>

#include "gazebo/Master.hh"
#include "gazebo/physics/physics.hh"
#include "gazebo/sensors/sensors.hh"
#include "gazebo/transport/transport.hh"
#include "gazebo/common/common.hh"
#include "gazebo/util/LogRecord.hh"
#include "gazebo/math/gzmath.hh"
#include "gazebo/gazebo_config.h"
#include "gazebo/gazebo.hh"

boost::mutex fini_mutex;
std::vector<gazebo::SystemPluginPtr> g_plugins;

gazebo::Master *g_master = NULL;

/////////////////////////////////////////////////
struct g_vectorStringDup
{
  char *operator()(const std::string &_s)
  {
    return strdup(_s.c_str());
  }
};

/////////////////////////////////////////////////
// This function is used by both setupClient and setupServer
bool setup(const std::string &_prefix, int _argc, char **_argv)
{
  gazebo::common::load();

  // The SDF find file callback.
  sdf::setFindCallback(boost::bind(&gazebo::common::find_file, _1));

  // Initialize the informational logger. This will log warnings, and
  // errors.
  gzLogInit(_prefix, "default.log");

  // Load all the system plugins
  for (std::vector<gazebo::SystemPluginPtr>::iterator iter =
       g_plugins.begin(); iter != g_plugins.end(); ++iter)
  {
    (*iter)->Load(_argc, _argv);
  }

  if (!gazebo::transport::init())
  {
    gzerr << "Unable to initialize transport.\n";
    return false;
  }

  // Make sure the model database has started.
  gazebo::common::ModelDatabase::Instance()->Start();

  // Run transport loop. Starts a thread
  gazebo::transport::run();

  // Init all system plugins
  for (std::vector<gazebo::SystemPluginPtr>::iterator iter = g_plugins.begin();
       iter != g_plugins.end(); ++iter)
  {
    (*iter)->Init();
  }

  return true;
}

/////////////////////////////////////////////////
void gazebo::printVersion()
{
  fprintf(stderr, "%s", GAZEBO_VERSION_HEADER);
}

/////////////////////////////////////////////////
void gazebo::addPlugin(const std::string &_filename)
{
  gazebo::addPlugin(_filename);
}

/////////////////////////////////////////////////
void gazebo::addPlugin(const std::string &_filename)
{
  if (_filename.empty())
    return;
  gazebo::SystemPluginPtr plugin =
    gazebo::SystemPlugin::Create(_filename, _filename);

  if (plugin)
  {
    if (plugin->GetType() != SYSTEM_PLUGIN)
    {
      gzerr << "System is attempting to load "
        << "a plugin, but detected an incorrect plugin type. "
        << "Plugin filename[" << _filename << "].\n";
      return;
    }
    g_plugins.push_back(plugin);
  }
}

/////////////////////////////////////////////////
bool gazebo::setupServer(int _argc, char **_argv)
{
  std::string host = "";
  unsigned int port = 0;

  gazebo::transport::get_master_uri(host, port);

  g_master = new gazebo::Master();
  g_master->Init(port);
  g_master->RunThread();

  if (!setup("server-", _argc, _argv))
  {
    gzerr << "Unable to setup Gazebo\n";
    return false;
  }

  if (!sensors::load())
  {
    gzerr << "Unable to load sensors\n";
    return false;
  }

  if (!gazebo::physics::load())
  {
    gzerr << "Unable to initialize physics.\n";
    return false;
  }

  if (!sensors::init())
  {
    gzerr << "Unable to initialize sensors\n";
    return false;
  }

  return true;
}

/////////////////////////////////////////////////
bool gazebo::setupServer(const std::vector<std::string> &_args)
{
  std::vector<char *> pointers(_args.size());
  std::transform(_args.begin(), _args.end(), pointers.begin(),
                 g_vectorStringDup());
  pointers.push_back(0);
  bool result = gazebo::setupServer(_args.size(), &pointers[0]);

  // Deallocate memory for the command line arguments allocated with strdup.
  for (size_t i = 0; i < pointers.size(); ++i)
    free(pointers.at(i));

  return result;
}

/////////////////////////////////////////////////
bool gazebo::setupClient(int _argc, char **_argv)
{
  if (!setup("client-", _argc, _argv))
  {
    gzerr << "Unable to setup Gazebo\n";
    return false;
  }

  common::Time waitTime(1, 0);
  int waitCount = 0;
  int maxWaitCount = 10;

  // Wait for namespaces.
  while (!gazebo::transport::waitForNamespaces(waitTime) &&
      (waitCount++) < maxWaitCount)
  {
    gzwarn << "Waited " << waitTime.Double() << "seconds for namespaces.\n";
  }

  if (waitCount >= maxWaitCount)
  {
    gzerr << "Waited " << (waitTime * waitCount).Double()
      << " seconds for namespaces. Giving up.\n";
  }

  return true;
}

/////////////////////////////////////////////////
bool gazebo::setupClient(const std::vector<std::string> &_args)
{
  std::vector<char *> pointers(_args.size());
  std::transform(_args.begin(), _args.end(), pointers.begin(),
                 g_vectorStringDup());
  pointers.push_back(0);
  bool result = gazebo::setupClient(_args.size(), &pointers[0]);

  // Deallocate memory for the command line arguments alloocated with strdup.
  for (size_t i = 0; i < pointers.size(); ++i)
    free(pointers.at(i));

  return result;
}

/////////////////////////////////////////////////
bool gazebo::shutdown()
{
  gazebo::physics::stop_worlds();

  gazebo::sensors::stop();

  // Stop log recording
  util::LogRecord::Instance()->Stop();

  // Stop transport
  gazebo::transport::stop();

  // Make sure to shut everything down.
  boost::mutex::scoped_lock lock(fini_mutex);
  util::LogRecord::Instance()->Fini();
  g_plugins.clear();
  gazebo::transport::fini();

  gazebo::physics::fini();

  gazebo::sensors::fini();

  delete g_master;
  g_master = NULL;

  // Cleanup model database.
  common::ModelDatabase::Instance()->Fini();

  return true;
}

/////////////////////////////////////////////////
gazebo::physics::WorldPtr gazebo::loadWorld(const std::string &_worldFile)
{
  gazebo::physics::WorldPtr world;

  // Load the world file
  sdf::SDFPtr sdf(new sdf::SDF);
  if (!sdf::init(sdf))
  {
    gzerr << "Unable to initialize sdf\n";
    return world;
  }

  // Find the file.
  std::string fullFile = gazebo::common::find_file(_worldFile);

  if (fullFile.empty())
  {
    gzerr << "Unable to find file[" << _worldFile << "]\n";
    return world;
  }

  if (!sdf::readFile(fullFile, sdf))
  {
    gzerr << "Unable to read sdf file[" << "empty.world" << "]\n";
    return world;
  }

  world = gazebo::physics::create_world();
  gazebo::physics::load_world(world, sdf->root->GetElement("world"));

  gazebo::physics::init_world(world);

  return world;
}

/////////////////////////////////////////////////
void gazebo::runWorld(gazebo::physics::WorldPtr _world,
                      unsigned int _iterations)
{
  if (!_world)
    gzerr << "World pointer is NULL\n";
  else
    _world->RunBlocking(_iterations);
}

/////////////////////////////////////////////////
bool gazebo::setupServer(int _argc, char **_argv)
{
  std::string host = "";
  unsigned int port = 0;

  gazebo::transport::get_master_uri(host, port);

  g_master = new gazebo::Master();
  g_master->Init(port);
  g_master->RunThread();

  if (!setup(_argc, _argv))
  {
    gzerr << "Unable to setup Gazebo\n";
    return false;
  }

  if (!sensors::load())
  {
    gzerr << "Unable to load sensors\n";
    return false;
  }

  if (!gazebo::physics::load())
  {
    gzerr << "Unable to initialize physics.\n";
    return false;
  }

  if (!sensors::init())
  {
    gzerr << "Unable to initialize sensors\n";
    return false;
  }

  return true;
}

/////////////////////////////////////////////////
bool gazebo::setupClient(int _argc, char **_argv)
{
  if (!setup(_argc, _argv))
  {
    gzerr << "Unable to setup Gazebo\n";
    return false;
  }

  common::Time waitTime(1, 0);
  int waitCount = 0;
  int maxWaitCount = 10;

  // Wait for namespaces.
  while (!gazebo::transport::waitForNamespaces(waitTime) &&
      (waitCount++) < maxWaitCount)
  {
    gzwarn << "Waited " << waitTime.Double() << "seconds for namespaces.\n";
  }

  if (waitCount >= maxWaitCount)
  {
    gzerr << "Waited " << (waitTime * waitCount).Double()
      << " seconds for namespaces. Giving up.\n";
  }

  return true;
}

/////////////////////////////////////////////////
bool gazebo::shutdown()
{
  gazebo::physics::stop_worlds();

  gazebo::sensors::stop();

  // Stop log recording
  util::LogRecord::Instance()->Stop();

  // Stop transport
  gazebo::transport::stop();

  // Make sure to shut everything down.
  boost::mutex::scoped_lock lock(fini_mutex);
  util::LogRecord::Instance()->Fini();
  g_plugins.clear();
  gazebo::transport::fini();

  gazebo::physics::fini();

  gazebo::sensors::fini();

  delete g_master;
  g_master = NULL;

  // Cleanup model database.
  common::ModelDatabase::Instance()->Fini();

  return true;
}

/////////////////////////////////////////////////
gazebo::physics::WorldPtr gazebo::loadWorld(const std::string &_worldFile)
{
  gazebo::physics::WorldPtr world;

  // Load the world file
  sdf::SDFPtr sdf(new sdf::SDF);
  if (!sdf::init(sdf))
  {
    gzerr << "Unable to initialize sdf\n";
    return world;
  }

  // Find the file.
  std::string fullFile = gazebo::common::find_file(_worldFile);

  if (fullFile.empty())
  {
    gzerr << "Unable to find file[" << _worldFile << "]\n";
    return world;
  }

  if (!sdf::readFile(fullFile, sdf))
  {
    gzerr << "Unable to read sdf file[" << "empty.world" << "]\n";
    return world;
  }

  world = gazebo::physics::create_world();
  gazebo::physics::load_world(world, sdf->root->GetElement("world"));

  gazebo::physics::init_world(world);

  return world;
}

/////////////////////////////////////////////////
void gazebo::runWorld(gazebo::physics::WorldPtr _world,
                      unsigned int _iterations)
{
  if (!_world)
    gzerr << "World pointer is NULL\n";
  else
    _world->RunBlocking(_iterations);
}
